Makefile-v2.mk

```diff
@@ -0,0 +1,296 @@
+##   GO_TAGS          Extra tags to use when building.
+
+# TODO(rfratto): test-packages target
+#
+# This depends on some reworking of how the Go tests in ./packaging works to
+# assume that packages have already been built so we don't have to hook in
+# packaging code here.
+
+include tools/make/*.mk
+
+AGENT_IMAGE     ?= grafana/agent:latest
+AGENTCTL_IMAGE  ?= grafana/agentctl:latest
+OPERATOR_IMAGE  ?= grafana/agent-operator:latest
+CROW_IMAGE      ?= us.gcr.io/kubernetes-dev/grafana/agent-crow:latest
+SMOKE_IMAGE     ?= us.gcr.io/kubernetes-dev/grafana/agent-smoke:latest
+AGENT_BINARY    ?= build/agent
+AGENTCTL_BINARY ?= build/agentctl
+OPERATOR_BINARY ?= build/agent-operator
+CROW_BINARY     ?= build/agent-crow
+SMOKE_BINARY    ?= build/agent-smoke
+GOOS            ?= $(shell go env GOOS)
+GOARCH          ?= $(shell go env GOARCH)
+GOARM           ?= $(shell go env GOARM)
+CGO_ENABLED     ?= 1
+RELEASE_BUILD   ?= 0
+
+# This should contain the list of all environment variables should should
+# propagate to the build container. USE_CONTAINER should _not_ be included to
+# avoid infinite recursion.
+PROPAGATE_VARS := \
+	AGENT_IMAGE AGENTCTL_IMAGE OPERATOR_IMAGE CROW_IMAGE SMOKE_IMAGE \
+	BUILD_IMAGE GOOS GOARCH GOARM CGO_ENABLED RELEASE_BUILD \
+	AGENT_BINARY AGENTCTL_BINARY OPERATOR_BINARY CROW_BINARY SMOKE_BINARY \
+	VERSION GO_TAGS
+
+#
+# Contants for targets
+#
+
+GO_ENV := GOOS=$(GOOS) GOARCH=$(GOARCH) GOARM=$(GOARM) CGO_ENABLED=$(CGO_ENABLED)
+
+VERSION      ?= $(shell ./tools/image-tag)
+GIT_REVISION := $(shell git rev-parse --short HEAD)
+GIT_BRANCH   := $(shell git rev-parse --abbrev-ref HEAD)
+VPREFIX      := github.com/grafana/agent/pkg/build
+GO_LDFLAGS   := -X $(VPREFIX).Branch=$(GIT_BRANCH)                        \
+                -X $(VPREFIX).Version=$(VERSION)                          \
+                -X $(VPREFIX).Revision=$(GIT_REVISION)                    \
+                -X $(VPREFIX).BuildUser=$(shell whoami)@$(shell hostname) \
+                -X $(VPREFIX).BuildDate=$(shell date -u +"%Y-%m-%dT%H:%M:%SZ")
+
+DEFAULT_FLAGS    := $(GO_FLAGS)
+DEBUG_GO_FLAGS   := -gcflags "all=-N -l" -ldflags "$(GO_LDFLAGS)" -tags "netgo $(GO_TAGS)"
+RELEASE_GO_FLAGS := -ldflags "-s -w $(GO_LDFLAGS)" -tags "netgo $(GO_TAGS)"
+
+ifeq ($(RELEASE_BUILD),1)
+GO_FLAGS := $(DEFAULT_FLAGS) $(RELEASE_GO_FLAGS)
+else
+GO_FLAGS := $(DEFAULT_FLAGS) $(DEBUG_GO_FLAGS)
+endif
+
+#
+# Targets for running tests
+#
+# These targets currently don't support proxying to a build container due to
+# difficulties with testing ./pkg/util/k8s and testing packages.
+#
+
+.PHONY: lint
+lint:
+	golangci-lint run -v --timeout=10m
+
+.PHONY: test
+# We have to run test twice: once for all packages with -race and then once
+# more without -race for packages that have known race detection issues.
+test:
+	$(GO_ENV) go test $(CGO_FLAGS) -race ./...
+	$(GO_ENV) go test $(CGO_FLAGS) ./pkg/integrations/node_exporter ./pkg/logs ./pkg/operator ./pkg/util/k8s
```

Comment:
I assume this step will succeed as long as the second go test is successful. How will we easily detect when a new package introduces race conditions?